### ISSUE TITLE
consult doesn't call initialization/1 directive goals

### ISSUE THREAD

**Issue by pmoura** [tau-prolog/tau-prolog#142]
the `consult` JavaScript function doesn't call `initialization/1` directive goals. For example, with:

```js
session.consult( ":- initialization(write(hello))." );
session.query( "true." );
session.answers( x => console.log( pl.format_answer(x) ) );
```

we get:

```bash
$ node ./tp.js 
true ;
false.
```

**Comment by pmoura** [comment]
There seems to be some issue with that commit:

```bash
$ node --stack_size=5000 ./tp.js
/Users/pmoura/Documents/Logtalk/tp/node_modules/tau-prolog/modules/core.js:1548
		this.session.total_threads++;
		     ^

TypeError: Cannot read property 'total_threads' of undefined
    at new Thread (/Users/pmoura/Documents/Logtalk/tp/node_modules/tau-prolog/modules/core.js:1548:8)
    at parseProgram (/Users/pmoura/Documents/Logtalk/tp/node_modules/tau-prolog/modules/core.js:1024:18)
    at Thread.consult (/Users/pmoura/Documents/Logtalk/tp/node_modules/tau-prolog/modules/core.js:2524:21)
    at Session.consult (/Users/pmoura/Documents/Logtalk/tp/node_modules/tau-prolog/modules/core.js:2420:22)
    at Object.<anonymous> (/Users/pmoura/Documents/Logtalk/tp/tp.js:12:9)
    at Module._compile (internal/modules/cjs/loader.js:1133:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1153:10)
    at Module.load (internal/modules/cjs/loader.js:977:32)
    at Function.Module._load (internal/modules/cjs/loader.js:877:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)
```

**Comment by pmoura** [comment]
To reproduce the bug:

```js
session.consult( ":- initialization(write(hello)). " );
session.query( "true." );
session.answers( x => console.log( pl.format_answer(x) ) );
```

**Comment by jariazavalverde** [comment]
Fixed!

**Comment by pmoura** [comment]
But there's another bug: `initialization/1` directives must be called only after a file is finished loading. In the current implementation, the initialization goals seem to being called as soon as the directives are parsed. The solution could be calling the initialization goals right before the query.

**Comment by jariazavalverde** [comment]
Initialization goals are stored into an array, and are called when the file is parsed:

```javascript
var session = pl.create();
session.consult(`
:- initialization((p(X), write(X))).
p(a).
`); // a
```

**Comment by pmoura** [comment]
But that's not standard behavior and it breaks existing code. The solution seems to be collecting all initialization goals from `session.consult()` calls and then call them when processing the `session.query()` call.

**Comment by pmoura** [comment]
For example, in my case I'm using:

```js
session.consult( "../logtalk3/adapters/unsupported/tau.pl" );
session.consult( "../logtalk3/paths/paths.pl" );
session.consult( "../logtalk3/core/core.pl" );
session.query( "true." );
session.answers( x => console.log( pl.format_answer(x) ) );
```

The `core.pl` file contains (as the last term) an `initialization/1` directive. I get:

```bash
$ node --stack_size=5000 ./tp.js
true ;
false.

Logtalk 3.39.0(-)b01
Copyright (c) 1998(-)2020 Paulo Moura
...
```

You can see above that the query is processed **before**  the initialization goal is called (which is what prints, among other tasks, the Logtalk banner).

**Comment by jariazavalverde** [comment]
I see. But waiting for the `query` method doesn't seem the most appropriate. What if `consult` method accepts a list of files, and the `initialization/1` directives are executed after the last file?

**Comment by pmoura** [comment]
A simple example:

```js
session.consult( ":- initialization((a(X), write(X)). a(hello)." );
session.consult( ":- initialization((b(X), write(X)). b(world)." );
session.query( "true." );
session.answers( x => console.log( pl.format_answer(x) ) );
```

We get:

```bash
$ node ./tp.js
true ;
false.
```

But the expected result would be:

```bash
$ node ./tp.js
helloworld
true ;
false.
```

**Comment by pmoura** [comment]
> I see. But waiting for the `query` method doesn't seem the most appropriate.

You're correct. The initialization goals must be called **after** the contents of the `session.consult()` call are fully parsed. I.e. after the "file" finished loading.

**Comment by jariazavalverde** [comment]
> The initialization goals for a given file must be called after the file is loaded. Thus, if we have e.g. a list of two files, each with its own set of `initialization/1` directives, the initialization goals for the first file must be called before parsing the second file and calling the second file initialization goals.

I am getting lost. That is what it currently does. Your example has a syntax error:

```javascript
session.consult( ":- initialization((a(X), write(X))). a(hello)." );
session.consult( ":- initialization((b(X), write(X))). b(world)." );
session.query( "true." );
session.answers( x => console.log( pl.format_answer(x) ) );
```

```
$ nodejs tau.js 
helloworldtrue ;
false.
```

**Comment by pmoura** [comment]
Oops! Sorry for the syntax error. But I'm still not getting the expected semantics on the big example. Debugging....

**Comment by pmoura** [comment]
... this is my code:

```js
var session = pl.create( 20000 );
session.consult( "../logtalk3/adapters/unsupported/tau.pl" );
session.consult( "../logtalk3/paths/paths.pl" );
session.consult( "../logtalk3/core/core.pl" );
session.query( "logtalk_load(hello_world(loader))." );
session.answers( x => console.log( pl.format_answer(x) ) );
```

As the `core.pl` file contains an `initialization/1` directive, which prints the banner and the default flags, I should that information printed **before** the query is run. The query is a call to load an example, that also contains `initialization/1` directives. But I get instead:

```bash
$ node --stack_size=5000 ./tp.js
core silent(loading): loading_file('/opt/local/share/logtalk/core/expanding.lgt', [code_prefix($), clean(off), scratch_directory('/Users/pmoura/logtalk/scratch/'), optimize(on), report(off), reload(skip)])
core silent(loading): loading_file('/Users/pmoura/logtalk/examples/hello_world/loader.lgt', [])
core silent(compiling): compiling_file('/Users/pmoura/logtalk/examples/hello_world/loader.lgt', [])
core silent(compiling): compiled_file('/Users/pmoura/logtalk/examples/hello_world/loader.lgt', [])
core silent(loading): loading_file('/Users/pmoura/logtalk/examples/hello_world/hello_world.lgt', [])
core silent(compiling): compiling_file('/Users/pmoura/logtalk/examples/hello_world/hello_world.lgt', [])
core silent(compiling): compiling_entity(object, hello_world)
core silent(compiling): compiled_entity(object, hello_world)
core silent(compiling): compiled_file('/Users/pmoura/logtalk/examples/hello_world/hello_world.lgt', [])
core comment(loading): loaded_file('/Users/pmoura/logtalk/examples/hello_world/loader.lgt', [])
true ;
Hello World!
false.
core comment(loading): loaded_file('/Users/pmoura/logtalk/examples/hello_world/hello_world.lgt', [])
core comment(warnings): compilation_and_loading_warnings(0, 0)
```

All the lines above starting with `core ...` would not be there if the initialization goal from the `core.pl` file was called.

**Comment by pmoura** [comment]
As we discussed elsewhere, the strange results above result from the initialization goals being handled asynchronously: the initialization goal in the `core.pl` file is still running when the initialization goal in the `hello_world.lgt` file starts to run. But the second initialization goal requires the first one to be completed. In general, if we're consult a list of files, each with its own initialization goal, these need to be completed before the next file is consulted in the list. Thus, it's not just initialization goals that need to be serialized but file consulting as well (if not already).

**Comment by jariazavalverde** [comment]
This is an important change in the consult process, since from now on the `consult` method is also an asynchronous process. `consult` accepts `success` option to be executed when the parsing is end, and  `error` option to report an error:

```javascript
var session = pl.create(20000);
var err = function(error) { console.log(pl.format_answer(error, session)); };
session.consult("../logtalk3/adapters/unsupported/tau.pl", {success: function() {
    session.consult("../logtalk3/paths/paths.pl", {success: function() {
        session.consult("../logtalk3/core/core.pl", {success: function() {
            session.query("logtalk_load(hello_world(loader)).");
            session.answers( x => console.log( pl.format_answer(x) ) );
        }, error: err});
    }, error: err});
}, error: err});
```

The `query` method will also require this change due to the goal expansion.

**Comment by pmoura** [comment]
How is this reflected in the `consult` **predicate**?

**Comment by jariazavalverde** [comment]
> How is this reflected in the `consult` **predicate**?

The `consult/1` built-in predicate should not be affected, since it was already asynchronous as it was incorporated into the Tau Prolog resolution logic.

**Comment by pmoura** [comment]
But that's the problem. Consult cannot be asynchronous as a file may depend on a file loaded before. My understanding of that JavaScript snippet is that it provides a solution to make consult asynchronous. But that doesn't seem to be possible when using the `consult/1` predicate.

**Comment by jariazavalverde** [comment]
> But that's the problem. Consult cannot be asynchronous as a file may depend on a file that must complete loading before. My understanding of that JavaScript snippet is that it provides a solution to make consult synchronous. But that doesn't seem to be possible when using the `consult/1` predicate.

`consult/1` is asynchronous in the sense that the Tau Prolgo thread "falls asleep" until the file has finished loading, and then resolution continues.

**Comment by pmoura** [comment]
Ok, let me do some experiments with your latest changes.

**Comment by pmoura** [comment]
Nothing happens when using your snippet:

```js
var session = pl.create( 20000 );
var show = x => console.log(pl.format_answer(x, session));
session.consult("../logtalk3/adapters/unsupported/tau.pl", {success: function() {
    session.consult("../logtalk3/paths/paths.pl", {success: function() {
        session.consult("../logtalk3/core/core.pl", {success: function() {
            session.query("logtalk_load(hello_world(loader)).");
            session.answers(show);
        }, error: show});
    }, error: show});
}, error: show});
```

I.e. the initialization goals are not called. The process exits almost immediately with no output.

**Comment by pmoura** [comment]
Also, all `consult/1` queries using the simple REPL code your posted in another issue now fail.

**Comment by jariazavalverde** [comment]
I tried this:

```javascript
var session = pl.create(20000);
var show = x => console.log(pl.format_answer(x, session));
session.consult("a.pl", {success: function() {
    session.consult("b.pl", {success: function() {
        session.consult("c.pl", {success: function() {
            session.query("write(done), nl.");
            session.answers(show);
        }, error: show});
    }, error: show});
}, error: show});
```

```shell
nodejs tau.js 
a
b
c
done
true.
false.
```

```prolog
% a.pl
:- initialization((p(X), write(X), nl)).
p(a).
```

```prolog
% b.pl
:- initialization((q(X), write(X), nl)).
q(b).
```

```prolog
% c.pl
:- initialization((r(X), write(X), nl)).
r(c).
```

**Comment by pmoura** [comment]
The other suggestion:

```js
var session = pl.create( 20000 );
session.query(`
consult('/Users/pmoura/Documents/Logtalk/logtalk3/adapters/unsupported/tau.pl').
`);
session.answers(x => console.log(pl.format_answer(x, session)));
```

just result in:

```bash
$ node --stack_size=5000 ./tp.js
false.
```

**Comment by pmoura** [comment]
> I tried this:

I can reproduce your experiment and I get the same results as you. But when I replace the paths to the `a.pl`, `b.pl`, and `c.pl` files with the paths to the files that I'm trying to consult, I simply immediately get:

```bash
$ node --stack_size=5000 ./tau.js
$
```

Tried both relative paths, which I used successfully before, and absolute paths. Same results.

**Comment by pmoura** [comment]
My suggestion for us to debug this issue is for you to clone the `logtalk3` repo and then reproduce my results using:

```js
var pl = require("./node_modules/tau-prolog/modules/core.js");
require("./node_modules/tau-prolog/modules/format.js")(pl);
require("./node_modules/tau-prolog/modules/js.js")(pl);
require("./node_modules/tau-prolog/modules/lists.js")(pl);
require("./node_modules/tau-prolog/modules/os.js")(pl);
require("./node_modules/tau-prolog/modules/random.js")(pl);
require("./node_modules/tau-prolog/modules/statistics.js")(pl);
//require("./node_modules/tau-prolog/modules/system.js")(pl);

var session = pl.create(20000);
var show = x => console.log(pl.format_answer(x, session));
session.consult("../logtalk3/adapters/unsupported/tau.pl", {success: function() {
    session.consult("../logtalk3/paths/paths.pl", {success: function() {
        session.consult("../logtalk3/core/core.pl", {success: function() {
            session.query("write(done), nl.");
            session.answers(show);
        }, error: show});
    }, error: show});
}, error: show});
```

You may need to adjust the paths.

**Comment by jariazavalverde** [comment]
Sorry, it was a bug with the `op/3` directive. I just tried logtalk and now it works correctly.

**Comment by pmoura** [comment]
Thanks for debugging the issue. I would never guessed a Tau bug in the `op/3` directive! I will update and resume testing.

**Comment by jariazavalverde** [comment]
This (d84181b) is the last major change in the Tau Prolog interface. `Thread.prototype.query` becomes asynchronous due to goal expansion.  @pmoura  Hopefully this is the last change for your script: :smile: 

```javascript
var readlineSync = require("readline-sync");
var pl = require("./node_modules/tau-prolog/modules/core.js");
require("./node_modules/tau-prolog/modules/format.js")(pl);
require("./node_modules/tau-prolog/modules/js.js")(pl);
require("./node_modules/tau-prolog/modules/lists.js")(pl);
require("./node_modules/tau-prolog/modules/os.js")(pl);
require("./node_modules/tau-prolog/modules/random.js")(pl);
require("./node_modules/tau-prolog/modules/statistics.js")(pl);

var max_answers = process.argv[2] || 100;
var session = pl.create(50000);
var compose = (f,g) => x => f(g(x));
var show = x => console.log(pl.format_answer(x, session));
var repl = function() {
    var query = readlineSync.question("?- ", {keepWhitespace: true}) + "\n";
    session.query(query, {success: function(_goal) {
        session.answers(function(answer) {
            show(answer);
            if(answer === false || pl.type.is_error(answer))
                repl();
        }, max_answers);
    }, error: compose(repl, show)});
};
session.consult("../logtalk3/adapters/unsupported/tau.pl", {success: function() {
    session.consult("../logtalk3/paths/paths.pl", {success: function() {
        session.consult("../logtalk3/core/core.pl", {success: function() {
            repl();
        }, error: show});
    }, error: show});
}, error: show});
```

**Comment by pmoura** [comment]
Thanks! I will use this update script going forward. I would like to include it in the Logtalk distribution. Can you suggest a comment header to give Tau Prolog proper credit? Logtalk itself is distributed under the Apache License 2.0.

**Comment by jariazavalverde** [comment]
Great! Please feel free to add the header that you consider appropriate. You can distribute it under the same Logtalk license.

**Comment by pmoura** [comment]
Done. You can browse the file at https://github.com/LogtalkDotOrg/logtalk3/blob/master/integration/logtalk_tau.js

The file is used by the new `taulgt` script (if you're running Logtalk from the `logtalk3` clone directory, you will need to use `taulgt.sh` instead after adding the `$LOGTALKHOME/integration` directory to your system `PATH`; see https://logtalk.org/running_developer_versions.html for details).

Experimenting with it, there's some strange interaction with multiple queries. For example:

```text
?- {ack(loader)}.
% [ (/)Users(/)pmoura(/)logtalk(/)examples(/)ack(/)ack.lgt loaded ]
% [ (/)Users(/)pmoura(/)logtalk(/)examples(/)ack(/)loader.lgt loaded ]
% (0 warnings)
true ;
false.
?- ack::ack(2, 4, V).
V = 11 ;
false.
?- {aliases(loader)}.
false.
```

At this point, if I simply press return, I get the output from loading the second example and then the process exits:

```text
?- 
(*)     Singleton variables(:) [Variables(,) Variables]
*       while compiling object square('$VAR'(0))
*       in file (/)Users(/)pmoura(/)logtalk(/)examples(/)aliases(/)aliases.lgt at or above line (-)1
*     
(*)     Singleton variables(:) [Variables(,) Variables]
*       while compiling object circle('$VAR'(0))
*       in file (/)Users(/)pmoura(/)logtalk(/)examples(/)aliases(/)aliases.lgt at or above line (-)1
*     
% [ (/)Users(/)pmoura(/)logtalk(/)examples(/)aliases(/)aliases.lgt loaded ]
% [ (/)Users(/)pmoura(/)logtalk(/)examples(/)aliases(/)loader.lgt loaded ]
% 2 compilation warnings
$
```

Maybe an issue with `readline-sync`?
